lms/djangoapps/instructor/services.py

```diff
@@ -78,13 +76,60 @@ def delete_student_attempt(self, student_identifier, course_id, content_id, requ
             except (StudentModule.DoesNotExist, enrollment.sub_api.SubmissionError):
                 err_msg = (
                     'Error occurred while attempting to reset student attempts for user '
-                    '{student_identifier} for content_id {content_id}.'.format(
-                        student_identifier=student_identifier,
-                        content_id=content_id
-                    )
+                    f'{student_identifier} for content_id {content_id}.'
                 )
                 log.error(err_msg)
 
+    def complete_student_attempt(self, user_identifier, content_id):
+        """
+        """
+        try:
+            user = get_user_by_username_or_email(user_identifier)
+        except ObjectDoesNotExist:
+            err_msg = (
+                'Error occurred while attempting to complete student attempt for user '
+                f'{user_identifier} for content_id {content_id}. '
+                'User does not exist!'
+            )
+            log.error(err_msg)
+            return
+
+        try:
+            block_key = UsageKey.from_string(content_id)
+        except InvalidKeyError:
+            err_msg = (
+                'Error occurred while attempting to complete student attempts for user '
+                f'{user_identifier} for content_id {content_id}. '
+                'Invalid content_id!'
+            )
+            log.error(err_msg)
+            return
+
+        def _submit_completions(block, user):
+            """
+            """
+            mode = XBlockCompletionMode.get_mode(block)
+            if mode == XBlockCompletionMode.COMPLETABLE:
+                block.runtime.publish(block, 'completion', {'completion': 1.0, 'user_id': user.id})
+            elif mode == XBlockCompletionMode.AGGREGATOR:
+                block_children = ((hasattr(block, 'get_child_descriptors') and block.get_child_descriptors())
+                                 or (hasattr(block, 'get_children') and block.get_children()))
+                for child in block_children:
+                    _submit_completions(child, user)
```

Comment:
not familiar with what XBlockCompletionMode.AGGREGATOR is, but wanted to check if there is any case where this could leave to an infinite recursive loop?

Reply:
I think technically if you manually edit your course xml, you can have infinite loops in your xblock structure. But just don't do that, pls. (And I'm sure other code breaks)

Reply:
Yeah, if you make a loop in your course, it will hit that, but if you do that, your entire course gonna be broke (I'd think) so not a big concern in my mind.